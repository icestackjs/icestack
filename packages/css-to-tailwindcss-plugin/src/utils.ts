import path from 'node:path'
import fs from 'node:fs'
export function resolvePath(p: string, cwd: string = process.cwd()) {
  return path.isAbsolute(p) ? p : path.resolve(cwd, p)
}

export function ensureDir(dir: string) {
  if (fs.existsSync(dir)) {
    const stat = fs.statSync(dir)
    if (!stat.isDirectory()) {
      fs.mkdirSync(dir)
    }
  } else {
    fs.mkdirSync(dir)
  }
}
