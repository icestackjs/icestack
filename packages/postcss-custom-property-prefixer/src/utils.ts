import { PrefixFunction, UserDefinedOptions } from './types'

export function matchCustomPropertyFromValue(str: string, cb: (arr: RegExpExecArray, index: number) => void) {
  let arr: RegExpExecArray | null
  let index = 0
  const regex = /var\(\s*(--\w[\w-]*)/g
  while ((arr = regex.exec(str)) !== null) {
    cb(arr, index)
    index++
  }
}

export function makeCustomProperty(customProperty: string, prefix: string) {
  return `--${prefix}${customProperty.slice(2)}`
}

export const postcssPlugin = 'postcss-custom-property-prefixer'

export function makePrefixFunction(prefix: UserDefinedOptions['prefix']): PrefixFunction {
  return typeof prefix === 'string'
    ? () => {
        return prefix
      }
    : prefix
}
