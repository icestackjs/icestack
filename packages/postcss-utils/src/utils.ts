import type { ChildNode } from 'postcss'
import sassParser from 'postcss-scss'
import selectorParser from 'postcss-selector-parser'
import { set, get } from 'lodash'
import type { CssValue, CssInJs } from '@icestack/types'

const defaultSelectorParser = selectorParser()

export function compressCssSelector(selectors: string) {
  return defaultSelectorParser.processSync(selectors, { lossless: false })
}

export function recursiveNodes(nodes: ChildNode[], result: Record<string, any> = {}) {
  for (const node of nodes) {
    switch (node.type) {
      case 'atrule': {
        if (node.name === 'apply') {
          const v = get(result, 'apply')
          if (Array.isArray(v)) {
            v.push(node.params)
          } else {
            set(result, 'apply', [node.params])
          }
        } else {
          const selector = `@${node.name} ${node.params}`
          result[selector] = {}
          recursiveNodes(node.nodes ?? [], result[selector])
        }

        break
      }
      case 'rule': {
        const s = compressCssSelector(node.selector)
        if (result[s] === undefined) {
          result[s] = {}
        }

        recursiveNodes(node.nodes, result[s])
        break
      }
      case 'decl': {
        set(result, `css.${node.prop}`, node.value + (node.important ? ' !important' : ''))
        break
      }
    }
  }
  return result
}

export function transformCss2Js(css: string): Record<string, any>
export function transformCss2Js(css: string | CssInJs): Record<string, any>
export function transformCss2Js<T>(css: T) {
  if (typeof css === 'string') {
    const root = sassParser.parse(css)
    const result = recursiveNodes(root.nodes)
    return result
  }
  return css
}

export function mapCss2JsArray(value?: CssValue) {
  if (value) {
    return Array.isArray(value) ? value.map((x) => transformCss2Js(x)) : [transformCss2Js(value)]
  }
  return []
}
