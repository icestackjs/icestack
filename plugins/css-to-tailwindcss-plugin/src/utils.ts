import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'

export function resolvePath(p: string, cwd: string = process.cwd()) {
  return path.isAbsolute(p) ? p : path.resolve(cwd, p)
}

export function ensureDir(dir: string) {
  if (fs.existsSync(dir)) {
    const stat = fs.statSync(dir)
    if (!stat.isDirectory()) {
      fs.mkdirSync(dir)
    }
  }
  else {
    fs.mkdirSync(dir)
  }
}
