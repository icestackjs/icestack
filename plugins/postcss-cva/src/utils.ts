import fs from 'node:fs'
import path from 'node:path'

export function ensureDir(dir: string) {
  try {
    fs.mkdirSync(dir, { recursive: true })
  } catch {}
}

export const matchAll = (regex: RegExp, str: string) => {
  const arr = []
  let res
  do {
    res = regex.exec(str)
    if (res) {
      arr.push(res)
    }
  } while (res !== null)
  return arr
}
