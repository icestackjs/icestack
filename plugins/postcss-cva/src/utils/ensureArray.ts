function isArray(arg: unknown): arg is any[] | readonly any[] {
  return Array.isArray(arg)
}

export default function ensureArray<T>(thing: readonly T[] | T | undefined | null): readonly T[] {
  if (isArray(thing)) return thing
  if (thing == null) return []
  return [thing]
}
