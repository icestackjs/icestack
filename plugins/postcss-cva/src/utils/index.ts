import fs from 'node:fs'

export function ensureDir(dir: string) {
  try {
    fs.mkdirSync(dir, { recursive: true })
  }
  catch {}
}

export function matchAll(regex: RegExp, str: string) {
  const arr = []
  let res
  do {
    res = regex.exec(str)
    if (res) {
      arr.push(res)
    }
  } while (res !== null)
  return arr
}
