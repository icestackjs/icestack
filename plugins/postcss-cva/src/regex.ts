import type { Comment, Rule } from 'postcss'
import parser from 'postcss-selector-parser'
import { trimStart } from 'lodash'
import { matchAll } from './utils'
import type { CommentType } from './types'

// b
export const baseRegex = new RegExp(/@b/.source, 'g')
// gb
export const defineBaseRegex = new RegExp(/@gb/.source, 'g')
// v
export const variantRegex = new RegExp(/@v/.source, 'g')
// gv
export const defineVariantRegex = new RegExp(/@gv/.source, 'g')
// cv
export const compoundVariantRegex = /@cv/g
// gcv
export const defineCompoundVariantRegex = /@gcv/g
// dv
export const defaultVariantRegex = new RegExp(/@dv/.source, 'g')
// meta
export const defineMetaRegex = new RegExp(/@meta/.source, 'g')

const regexArray: { type: CommentType; regex: RegExp; next: boolean }[] = [
  {
    regex: baseRegex,
    type: 'base',
    next: true
  },
  {
    regex: variantRegex,
    type: 'variant',
    next: true
  },
  {
    regex: compoundVariantRegex,
    type: 'compoundVariant',
    next: true
  },
  {
    regex: defineBaseRegex,
    type: 'base',
    next: false
  },
  {
    regex: defineVariantRegex,
    type: 'variant',
    next: false
  },
  {
    regex: defineCompoundVariantRegex,
    type: 'compoundVariant',
    next: false
  },
  {
    regex: defaultVariantRegex,
    type: 'defaultVariant',
    next: false
  },
  {
    regex: defineMetaRegex,
    type: 'meta',
    next: false
  }
]

export function getSuffix(text: string) {
  for (const { next, regex, type } of regexArray) {
    regex.lastIndex = 0
    const res = regex.test(text)
    if (res) {
      return {
        type,
        suffix: text.slice(regex.lastIndex),
        next
      }
    }
  }
}

export function pickComment(comment: Comment) {
  const text = comment.text
  const p = text.indexOf('@')
  if (p > -1) {
    return getSuffix(text)
  }
}

export function extractParams(text: string, opts?: ReturnType<typeof pickComment>) {
  const { type } = opts ?? {}
  const params: string[] = []

  const query: Record<
    string,
    {
      value: string
    }
  > = {}

  function handle(res: string) {
    return type === 'meta' ? res : trimStart(res, '.')
  }

  const paramsArray = matchAll(/\[([^\]]*)]/g, text)
  for (const x of paramsArray) {
    const arr = x[1]
      .split(',')
      .map((x) => x.trim())
      .filter(Boolean)
    for (const d of arr) {
      if (d[0] === '"' && d.at(-1) === '"') {
        const param = handle(d.slice(1, -1))
        params.push(param)
      }
    }
  }
  const queryArray = matchAll(/([\w-]+)="([^"]+)"/g, text)

  for (const x of queryArray) {
    // const key = x[1]
    // const d = x[2]

    // if (d[0] === '"' && d.at(-1) === '"') {
    //   query[key] = {
    //     value: trimStart(d.slice(1, -1), '.')
    //   }
    // }
    const key = x[1]
    const d = x[2]
    query[key] = {
      value: handle(d)
    }
  }

  return {
    query,
    params
  }
}

export const defaultParser = parser()

export function getParentRule(comment: Comment) {
  const p = comment.parent
  if (p?.type === 'rule') {
    return p as Rule
  }
}

export const cvaSymbol = Symbol('cva')

export function setAdd<T>(set: Set<T>, value: T | T[]) {
  if (Array.isArray(value)) {
    for (const v of value) {
      set.add(v)
    }
  } else {
    set.add(value)
  }
}
