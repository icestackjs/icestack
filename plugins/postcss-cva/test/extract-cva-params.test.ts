import postcss, { comment } from 'postcss'
import extractCva from '@/extract'
import { baseRegex, compoundVariantRegex, defaultVariantRegex, variantRegex, pickComment, extractParams } from '@/regex'
import { matchAll } from '@/utils'

describe('extract-cva-params', () => {
  describe('baseRegex', () => {
    it('baseRegex case 0', () => {
      let res = matchAll(baseRegex, '@base=xxx')
      expect(res.length).toBe(1)

      res = matchAll(baseRegex, '@b=xxx')
      expect(res.length).toBe(1)

      res = matchAll(baseRegex, '@b "xxx"')
      expect(res.length).toBe(1)
    })
  })

  describe('variantRegex', () => {
    it('variantRegex case 0', () => {
      let res = matchAll(variantRegex, '@v ssd=true')
      expect(res.length).toBe(1)

      res = matchAll(variantRegex, '@variant dsa="xxx"')
      expect(res.length).toBe(1)

      res = matchAll(variantRegex, '@v ssd')
      expect(res.length).toBe(1)

      // res = matchAll(variantRegex, '@b xxx')
      // expect(res.length).toBe(1)
      // expect(res[0][1]).toBe(undefined)
    })
  })

  describe('defaultVariantRegex', () => {
    it('defaultVariantRegex case 0', () => {
      let res = matchAll(defaultVariantRegex, '@dv ssd=true')
      expect(res.length).toBe(1)

      res = matchAll(defaultVariantRegex, '@dv dsa="xxx"')
      expect(res.length).toBe(1)

      res = matchAll(defaultVariantRegex, '@dv ssd')
      expect(res.length).toBe(1)

      // res = matchAll(variantRegex, '@b xxx')
      // expect(res.length).toBe(1)
      // expect(res[0][1]).toBe(undefined)
    })
  })

  describe('pickComment', () => {
    it('pickComment base case 0', () => {
      const res = pickComment(
        comment({
          text: '@b v'
        })
      )
      expect(res?.type).toBe('base')
      expect(res?.next).toBe(true)
      expect(res?.suffix).toBe(' v')
    })
  })

  describe('extractParams', () => {
    it('extractParams case 0', () => {
      let { query: res } = extractParams(' xxx="xx"  yyy="xxx" ')
      expect(res.xxx.value).toBe('xx')
      expect(res.yyy.value).toBe('xxx')
      const { query } = extractParams(' xxx=this  yyy="xxx" ')
      res = query
      expect(res.xxx).toBe(undefined)
      expect(res.yyy.value).toBe('xxx')
      // expect(res.xxx.this).toBe(true)
    })
  })

  it('base', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      
      .btn{
        /* @base */
        /* @base */
        color:red;
      }

      .btn{
        /* @base */
        /* @base */
        color:red;
      }
    `)

    expect(res).toMatchSnapshot()
  })

  it('define base', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      
      .btn{
        /* @gbase ["xx","yyy"] */
        color:red;
      }

      .btn{
        /* @gb ["yy","yyy"] */
        color:red;
      }
    `)

    expect(res).toMatchSnapshot()
  })

  it('v case1', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      
      .btn{
        /* @v intent="primary" */
        color:red;
      }

      
      .btn-primary{
        /* @v intent="primary" */
        color:red;
      }

    `)

    expect(res).toMatchSnapshot()
  })

  it('gv case1', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      
      .btn{
        /* @gv intent="primary" ["btn"] */
        color:red;
      }

      
      .btn-primary{
        /* @gv intent="primary" ["btn","btn-primary"] */
        color:red;
      }

    `)

    expect(res).toMatchSnapshot()
  })

  it('cv case1', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      
      .bbb{
        /* @cv intent="primary" */
      }

      
      .ccc{
        /* @cv intent="primary" */
      }

    `)

    expect(res).toMatchSnapshot()
  })

  it('gcv case1', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      
      .bbb{
        /* @gcv intent="primary" ["bbb","ccc"] */
      }

      
      .ccc{
        /* @gcv intent="primary" ["aaa","bbb"] */
      }

    `)

    expect(res).toMatchSnapshot()
  })
  it('dv case1', async () => {
    let res
    await postcss([
      extractCva({
        process(x) {
          res = x
        }
      })
      // @ts-ignore
    ]).process(`
      /* @dv intent="primary" */
     

    `)

    expect(res).toMatchSnapshot()
  })

  it('@meta case0', () => {
    const res = extractParams(' path="./nest-btn-4"')

    expect(res).toMatchSnapshot()
  })

  it('@meta case1', () => {
    const res = extractParams(' path="./nest-btn-4"', {
      next: false,
      suffix: '',
      type: 'meta'
    })

    expect(res).toMatchSnapshot()
  })
})
