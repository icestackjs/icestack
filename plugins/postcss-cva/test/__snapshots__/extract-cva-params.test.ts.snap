// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract-cva-params > @meta case0 1`] = `
{
  "params": [],
  "query": {
    "path": {
      "value": "/nest-btn-4",
    },
  },
}
`;

exports[`extract-cva-params > @meta case1 1`] = `
{
  "params": [],
  "query": {
    "path": {
      "value": "./nest-btn-4",
    },
  },
}
`;

exports[`extract-cva-params > base 1`] = `
{
  "base": [
    "btn",
  ],
  "compoundVariants": [],
  "defaultVariants": {},
  "file": undefined,
  "meta": {},
  "variants": {},
}
`;

exports[`extract-cva-params > cv case1 1`] = `
{
  "base": [],
  "compoundVariants": [
    {
      "class": [
        "bbb",
        "ccc",
      ],
      "intent": "primary",
    },
  ],
  "defaultVariants": {},
  "file": undefined,
  "meta": {},
  "variants": {},
}
`;

exports[`extract-cva-params > define base 1`] = `
{
  "base": [
    "xx",
    "yyy",
    "yy",
  ],
  "compoundVariants": [],
  "defaultVariants": {},
  "file": undefined,
  "meta": {},
  "variants": {},
}
`;

exports[`extract-cva-params > dv case1 1`] = `
{
  "base": [],
  "compoundVariants": [],
  "defaultVariants": {
    "intent": "primary",
  },
  "file": undefined,
  "meta": {},
  "variants": {},
}
`;

exports[`extract-cva-params > gcv case1 1`] = `
{
  "base": [],
  "compoundVariants": [
    {
      "class": [
        "bbb",
        "ccc",
        "aaa",
      ],
      "intent": "primary",
    },
  ],
  "defaultVariants": {},
  "file": undefined,
  "meta": {},
  "variants": {},
}
`;

exports[`extract-cva-params > gv case1 1`] = `
{
  "base": [],
  "compoundVariants": [],
  "defaultVariants": {},
  "file": undefined,
  "meta": {},
  "variants": {
    "intent": {
      "primary": [
        "btn",
        "btn-primary",
      ],
    },
  },
}
`;

exports[`extract-cva-params > v case1 1`] = `
{
  "base": [],
  "compoundVariants": [],
  "defaultVariants": {},
  "file": undefined,
  "meta": {},
  "variants": {
    "intent": {
      "primary": [
        "btn",
        "btn-primary",
      ],
    },
  },
}
`;
